pkg/operator/ceph/cluster/cluster.go

```diff
@@ -407,3 +419,94 @@ func diffImageSpecAndClusterRunningVersion(imageSpecVersion cephver.CephVersion,
 
 	return false, nil
 }
+
+// postMonStartupActions is a collection of actions to run once the monitors are up and running
+// It gets executed right after the main mon Start() method
+// Basically, it is executed between the monitors and the manager sequence
+func (c *cluster) postMonStartupActions() error {
+	// Create CSI Ceph key
+	csiSecretKey, err := createCSIKeyring(c.context, c.Namespace, &c.ownerRef)
+	if err != nil {
+		return fmt.Errorf("failed to create csi ceph keyring. %+v", err)
+	}
+
+	// Create or update Kubernetes CSI secret
+	if err := createOrUpdateCSISecret(c.context, c.Namespace, csiSecretKey, c.Info, &c.ownerRef); err != nil {
+		return fmt.Errorf("failed to create kubernetes csi secret. %+v", err)
+	}
+
+	// Enable Ceph messenger 2 protocol on Nautilus
+	if c.Info.CephVersion.IsAtLeastNautilus() {
+		v, err := client.GetCephMonVersion(c.context, c.Info.Name)
+		if err != nil {
+			return fmt.Errorf("failed to get ceph mon version. %+v", err)
+		}
+		if v.IsAtLeastNautilus() {
+			versions, err := client.GetAllCephDaemonVersions(c.context, c.Info.Name)
+			if err != nil {
+				return fmt.Errorf("failed to get ceph daemons versions. %+v", err)
+			}
+			if len(versions.Mon) == 1 {
+				// If length is one, this clearly indicates that all the mons are running the same version
+				// We are doing this because 'ceph version' might return the Ceph version that a majority of mons has but not all of them
+				// so instead of trying to active msgr2 when mons are not ready, we activate it when we believe that's the right time
+				client.EnableMessenger2(c.context, c.Namespace)
+			}
+		}
+	}
+
+	return nil
+}
+
+func createCSIKeyring(context *clusterd.Context, clusterName string, ownerRef *metav1.OwnerReference) (string, error) {
+	s := keyring.GetSecretStore(context, clusterName, ownerRef)
+
+	key, err := s.GenerateKey(csiUsername, cephCSIKeyringCaps())
+	if err != nil {
+		return "", err
+	}
+
+	return key, nil
+}
+
+func cephCSIKeyringCaps() []string {
```

Comment:
as this is specific to rbd can we rename this function because we are need to add similar function for cephfs